README.md
# associative-search-network


ChangeLog.md
# Changelog for associative-search-network

## Unreleased changes


src/AdaptiveElement.hs
module AdaptiveElement where

import Control.Applicative
import Control.Monad
import Control.Monad.Random
import Control.Monad.Reader
import Control.Arrow
import qualified System.Random.MWC.Distributions as Dist
import qualified System.Random.MWC as MWC

import Debug.Trace (trace)

type Weight = Double
type Input = Double
type InTrace = Double
type AdaptiveElement = [Weight]

data Env = Env { mean :: Double
               , stDev :: Double
               , alfa :: Double
               , beta :: Double
               , lr :: Double
               , signals :: [(SignalGen, Weight -> Weight)]
               }
data Signal = ON | OFF
type TimeStamp = Int
type SignalGen = TimeStamp -> Signal
type Stack g a = RandT g (Reader Env) a

positive :: Double -> Double
positive v
  | v > 0.0 = 1.0
  | otherwise = 0.0

noise :: RandomGen g => Input -> Stack g Double
noise x = do
  (m, sd) <- asks $ mean &&& stDev
  n <- getRandomR (m - sd, m + sd)
  return (x - n)

output :: RandomGen g => [Weight] -> [Input] -> Stack g Double
output ws xs = fmap positive  (sum ws xs $ pure 0.0)
  where
    sum :: [Weight] -> [Input] -> Stack g Double -> Stack g Double
    sum [] [] ms = ms
    sum (w:ws) (x:xs) ms = do
      s <- ms
      nx <- noise x
      sum ws xs $ pure (s + w*nx)

inputTrace :: RandomGen g => (InTrace, Input) -> Stack g Double
inputTrace (xt, x) = do
  c <- asks lr
  return $ c*xt + x

toDouble :: Signal -> Double
toDouble ON = 1.0
toDouble OFF = 0.0

prev :: TimeStamp -> TimeStamp
prev t = t - 1

next :: TimeStamp -> TimeStamp
next t = t + 1

weights :: RandomGen g => [Weight] -> [InTrace] -> Double -> Double -> Stack g [Weight]
weights ws xts resDiff yDiff = do
  (c, fs) <- asks $ lr &&& fmap snd . signals
  let
    loop :: [Weight] -> [InTrace] -> [Weight -> Weight] -> [Weight]
    loop [] [] [] = []
    loop (w:ws) (xt:xts) (f:fs) = f (w + c*resDiff*yDiff*xt) : loop ws xts fs
  return $ loop ws xts fs

repl :: RandomGen g => a -> Stack g [a]
repl a = do 
  sgs <- asks signals
  return $ replicate (length sgs) a

initWeights :: RandomGen g => Stack g [Weight]
initWeights = repl 0.0

initInTraces :: RandomGen g => Stack g [InTrace]
initInTraces = repl 0.0

adapt :: RandomGen g => ((TimeStamp, [Weight]) -> Bool) -> Stack g [Weight]
adapt f = do
  ws <- initWeights
  xts <- initInTraces
  loop 0 ws xts 0.0 0.0 0.0
  where
    inputs :: TimeStamp -> Stack g [Input]
    inputs t = do
      sgs <- asks signals
      return $ fmap (toDouble . ($ t) . fst) sgs
    loop :: TimeStamp -> [Weight] -> [InTrace] -> Double -> Double -> Double -> Stack g [Weight]
    loop t ws xts y y' y'' = do
      xs <- inputs t
      ws' <- weights ws xts 1.0 (y''-y')
      ny <- output ws xs
      xs <- inputs (t-1)
      xts' <- sequence (inputTrace <$> zip xts xs)
      if f (t,ws') then
        return ws'
      else
        loop (t+1) ws' xts' ny y y'

untilTs :: TimeStamp -> (TimeStamp, [Weight]) -> Bool
untilTs t (t', _) = t == t'

signalGen :: Int -> Int -> Int -> TimeStamp -> Signal
signalGen delay d td t = if t' >= delay && t' <= delay+(d-1) then ON else OFF
  where t' = t `mod` td

fig12 :: IO [Weight]
fig12 = MWC.withSystemRandom (\g -> runReader (evalRandT (adapt $ untilTs (td*100)) g) env)
  where
    env = Env { mean = 0.005
              , stDev = 0.3
              , alfa = 0.9
              , beta = 0.0
              , lr = 0.2
              , signals = outSigGen:inSigGen
              }
    td = 3+3+20+10
    inSigGen = replicate 4 (signalGen 0 3 td, id)
    outSigGen = (signalGen 3 3 td, const 0.6)    
    
debug :: Show a => a -> a
debug a | trace (show a) False = undefined
debug a = a

src/AssociativeSearchNetwork.hs
{-# LANGUAGE TupleSections #-}
module AssociativeSearchNetwork ( fig12
                                , Weight
                                , TimeStamp
                                , fig4
                                ) where

import Control.Applicative
import Control.Monad
import Control.Monad.Random
import Control.Monad.Reader
import Control.Arrow
import System.Random
import Text.Printf
import Debug.Trace (trace)

type Weight = Double
type Input = Double
type InTrace = Double
type Output = Double

data Configs = C { mean :: !Double
                 , stDev :: !Double
                 , distribution :: ![(Double, Rational)]
                 , alfa :: !Double
                 , signals :: ![(SignalGen, Weight -> Weight)]
                 , environment :: TimeStamp -> [Double] -> Double
                 }
data AdaptiveElement = AE { weights :: ![Weight]
                          , inputTraces :: ![InTrace]
                          , lRate :: TimeStamp -> Double
                          , output :: !Double
                          , prevOutput :: !Double
                          , prevPrevOutput :: !Double
                          }

data Signal = ON | OFF deriving Show
type TimeStamp = Int
type SignalGen = TimeStamp -> Signal
type Stack g a = RandT g (Reader Configs) a

printD :: Double -> String 
printD = printf "%.4f"

debug :: [Double] -> String -> [Double]
debug xs l | trace (l ++ foldl (\s d -> s ++ printD d ++ ", ") "[" xs ++ "]") False = undefined
debug xs l = xs

debugD :: Double -> String -> Double
debugD d l | trace (l ++ printD d) False = undefined
debugD d l = d

debugS :: Show a => String ->  a -> a
debugS l d | trace (l ++ show d) False = undefined
debugS l d = d

positive :: Double -> Double
positive v
  | v > 0.0 = 1.0
  | otherwise = 0.0

normal :: Double -> Double -> [(Double, Rational)]
normal m sd = loop 1 $ negate r
  where
    r = sd
    step :: Int -> Int
    step n = div n 50
    loop :: Int -> Double -> [(Double, Rational)]
    loop n v
      | v <= m = (v, toRational n) : loop (n + step n) (v + 0.001)
      | v > m && v <= r = (v, toRational n) : loop (n - step n) (v + 0.001)
      | v > r = []

noise :: RandomGen g => Double -> Stack g Double
noise x = do
  d <- asks distribution
  n <- fromList d
  return $ x + n

sumProds :: [Double] -> [Double] -> Double
sumProds xs ys = loop xs ys 0.0
  where
    loop :: [Double] -> [Double] -> Double ->  Double
    loop [] [] s = s
    loop (w:ws) (x:xs) s = loop ws xs (s + w*x)

calcOutput :: RandomGen g => [Input] -> [Weight] -> Stack g Double
calcOutput xs ws = do
  ws' <- sequence $ fmap noise ws
  return . positive $ sumProds ws' xs

calcInputTraces :: [Input] -> [InTrace] -> Stack g [Double]
calcInputTraces xs xts = do
  a <- asks alfa
  let
    loop :: [InTrace] -> [Input] -> [Double]
    loop [] [] = []
    loop (xt:xts) (x:xs) = (a*xt + x) : loop xts xs
  return $ loop xts xs

toDouble :: Signal -> Double
toDouble ON = 1.0
toDouble OFF = 0.0

repl :: a -> Stack g [a]
repl a = do 
  sgs <- asks signals
  return $ replicate (length sgs) a

inputs :: TimeStamp -> Stack g [Input]
inputs t = do
  sgs <- asks signals
  return $ fmap (toDouble . ($ t) . fst) sgs

tl :: String -> TimeStamp -> String
tl pre t = pre ++ "(" ++ show t ++ ") = "

adaptOutput :: RandomGen g => TimeStamp -> AdaptiveElement -> Stack g AdaptiveElement
adaptOutput t ae = do
  xs <- inputs (t-1)
  xts <- calcInputTraces xs $ inputTraces ae
  y <- calcOutput (debug xs $ tl "xs" (t-1)) $ weights ae
  return AE { weights = weights ae
            , inputTraces = xts
            , output = y
            , prevOutput = output ae
            , prevPrevOutput = prevOutput ae
            , lRate = lRate ae
            }

adaptWeight :: Double -> TimeStamp -> AdaptiveElement -> Stack g AdaptiveElement
adaptWeight zDiff t ae = do
  fs <- asks $ fmap snd . signals
  let
    yDiff = prevOutput ae - prevPrevOutput ae
    loop :: [Weight] -> [InTrace] -> [Weight -> Weight] -> [Weight]
    loop [] [] [] = []
    loop (w:ws) (xt:xts) (f:fs) = f (w + lRate ae t *(debugD zDiff "zDiff = ")*(debugD yDiff "yDiff = ")*xt) : loop ws xts fs
    ws = loop (weights ae) (inputTraces ae) fs
  return AE { weights = debug ws $ tl "ws" t
            , inputTraces = inputTraces ae
            , output = output ae
            , prevOutput = prevOutput ae
            , prevPrevOutput = prevPrevOutput ae
            , lRate = lRate ae
            }

asn :: Configs -> [AdaptiveElement] -> ((TimeStamp, [AdaptiveElement]) -> Bool) -> IO ([(TimeStamp, Double)], [AdaptiveElement])
asn cs aes until = do
  gen <- newStdGen
  return $ runReader (flip evalRandT gen $ search 0 [] aes) cs
    where
      search :: RandomGen g => TimeStamp -> [(TimeStamp, Double)] -> [AdaptiveElement] -> Stack g ([(TimeStamp, Double)], [AdaptiveElement])
      search t ts aes= do
        env <- asks environment
        aes' <- forM aes $ adaptOutput t
        let 
          z = env t (fmap prevOutput aes')
          zDiff = z - env (t-1) (fmap prevPrevOutput aes')
          ts' = (t, z):ts
        aes'' <- forM aes' $ adaptWeight zDiff t
        if until (t, aes'') then
          return (ts', aes'')
        else
          search (t+1) ts' aes''

initAE :: Int -> (TimeStamp -> Double) -> AdaptiveElement
initAE n lr = 
  let ds = replicate n 0.0
  in AE { weights = ds
        , inputTraces = ds
        , lRate = lr
        , output = 0.0
        , prevOutput = 0.0
        , prevPrevOutput = 0.0
        }

asnReset :: Int -> [SignalGen] -> (TimeStamp -> [Output] -> Double) -> IO [(TimeStamp, Double)]
asnReset n sgs env = fmap fst $ asn cs aes $ untilTS 200
  where
    cs = C { mean = 0.0
           , stDev = 0.1
           , distribution = normal 0.0 0.1
           , alfa = 0.0
           , signals = fmap (,id) sgs
           , environment = env
           }
    resetLr t
      | t `mod` 10 == 0 = 0.0
      | otherwise = 0.03
    aes = replicate n $ initAE (length sgs) resetLr

contextSwitch :: a -> a -> a -> TimeStamp -> a
contextSwitch v1 v2 d t
  | t < 0 = d
  | div t 10 `mod` 2 == 0 = v1
  | otherwise = v2

fig4 :: IO [(TimeStamp, Double)]
fig4 = asnReset 9 signalSwitches env
  where
    env t ys =
      let z = sumProds (debug ys $ tl "ys" t)
      in contextSwitch (z (debug [-1,1,1,1,-1,1,1,1,-1] $ tl "ya" t)) (z (debug [1,1,-1,-1,1,-1,-1,1,1] $ tl "ya" t)) 0 t
    xs = replicate 4 ON ++ replicate 4 OFF
    xs' = reverse xs
    signalSwitch (x,x') = contextSwitch x x' OFF
    signalSwitches = signalSwitch <$> zip xs xs'

{-asnWithPred :: Int 
asnWithPred n = asn cs 
  where
    cs = C { mean = 0.0
           , stDev = 0.1
           , alfa = 0.0
          , signals = sgs
           , 
           }
    p = initAE 0.1
    aes = p : replicate n $ initAE $ const 0.03
-}

untilTS :: TimeStamp -> (TimeStamp, [AdaptiveElement]) -> Bool
untilTS t (t', _) = t == t'

untilAvgW :: Weight -> (TimeStamp, [AdaptiveElement]) -> Bool
untilAvgW w (_, aes) =  sum ws / fromIntegral (length ws) >= w
  where ws = join $ fmap weights aes

signalGen :: Int -> Int -> Int -> TimeStamp -> Signal
signalGen delay d td t = if t >= 0 && t' >= delay && t' <= delay+(d-1) then ON else OFF
  where t' = t `mod` td

fig12 :: IO (TimeStamp, [Weight])
fig12 = do
  (ts, ws) <- asn cs [initAE (length sgs) $ const 0.2] $ untilAvgW 0.6
  return (fst $ head ts, head $ fmap weights ws)
  where
    td = 3+20+27
    inSigGen = replicate 4 (signalGen 0 3 td, id)
    outSigGen = (signalGen 3 20 td, const 0.6)
    sgs = outSigGen:inSigGen
    cs = C { mean = 0.005
           , stDev = 0.03
           , distribution = normal 0.005 0.03
           , alfa = 0.9
           , signals = sgs
           , environment = const $ const 1.0
           }